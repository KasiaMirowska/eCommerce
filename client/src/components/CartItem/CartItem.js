import React from 'react';
import './CartItem.styles.scss';


const CartItem = (props) => {
    const {imageUrl, price, name, quantity} = props;
    
    return (
        <div className='cart-item'>
            <img src={imageUrl} alt='item'/>
            <div className='item-details'>
                <span>{name}</span>
                <span>{quantity} x ${price}</span>
            </div>
        </div>
    )
  
}

export default React.memo(CartItem);