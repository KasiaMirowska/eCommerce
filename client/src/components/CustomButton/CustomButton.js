import React from 'react';
//import './CustomButton.styles.scss';
import {CustomButtonContaner} from './CustomButton.styles';

// export default function CustomButton({children, isGoogleSignedIn, invertedColor, ...otherProps}) {
//     //conditionally rendering a button class based on a prop for google signIn...if google? then it has a google class , if not then nothing plus always custom-button class...

//     return (
//         <button className={`${isGoogleSignedIn? 'google-sign-in' : ''} ${invertedColor? 'invertedColor' : ''} custom-button`} {...otherProps}>
//             {children}
//         </button>
//     )
// }


export default function CustomButton(props) {
    //after creating styled component this becomes simpler then above - we put rendering conditionals into the styled component

    return (
        <CustomButtonContaner {...props} >
            {props.children}
        </CustomButtonContaner>
    )
}