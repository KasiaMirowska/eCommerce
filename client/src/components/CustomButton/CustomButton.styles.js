import styled, {css} from 'styled-components';

//instead of conditionally rendering classes in the component to implement different rendering needs we will condionally render css bloks here inside the styled component
const getButtonStyles = props => {
    if(props.googleSignInS) {
        return googleSignInStyles;
    } 
    return props.inverted? invertedButtonStyles : buttonStyles;
}


const invertedButtonStyles = css `
    background-color: white;
    color: black;

    &:hover {
    background-color:black;
    color: white;
    }
`

const googleSignInStyles = css `
    background-color: #4285f4;
    color: white;

    &:hover {
        background-color:#357ae8;
        border: none;
    }
`

const buttonStyles = css `
    background-color: black;
    color: white;

    &:hover {
        background-color: rgb(210, 250, 81);
        color: black;
        }
`
export const CustomButtonContaner = styled.button `
    min-width: 165px;
    width: auto;
    height: 50px;
    letter-spacing: 0.5px;
    line-height: 50px;
    padding: 0 35px 0 35px;
    font-size: 15px;
    text-transform: uppercase;
    font-family: 'Open Sans Condensed';
    font-weight: bolder;
    border: none;
    cursor: pointer;
    display:flex;
    justify-content: center;

    ${getButtonStyles} 
    // calling the above function determining which style to render

`