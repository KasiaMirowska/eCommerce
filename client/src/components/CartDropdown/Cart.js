// import React from 'react';
// import CustomButton from '../CustomButton/CustomButton';
// import CartItem from '../CartItem/CartItem';
// import './Cart.styles.scss';
// import { connect } from 'react-redux';
// import { selectCartItems } from '../../redux/cart/cart.selectors';
// // import { selectCartHidden } from '../../redux/cart/cart.selectors';
// import { createStructuredSelector } from 'reselect';
// import { withRouter } from 'react-router-dom';
// import { toggleCartHidden } from '../../redux/cart/cart.actions';

// const mapStateToProps = createStructuredSelector({
//     // cartItems: state.cart.cartItems,
//     cartItems: selectCartItems,
// })

// export const Cart = (props) => {
//     //connect passes dispatch as a prop to component even if we do not provide a second argument to connect, see:
//     console.log(props, 'IN CART')
//     const items = props.cartItems.map(item => (
//         <CartItem key={item.id} {...item} />
//     ));

//     return (
//         <div className='cart-dropdown'>
//             <div className='cart-items'>
//                 {props.cartItems.length > 0 ?
//                     items
//                     :
//                     <p className='empty-message'>Your cart is empty</p>}
//                 <CustomButton onClick={() => {
//                     props.history.push('/checkout');
//                     props.dispatch(toggleCartHidden())
//                 }}
//                 >
//                     GO TO CHECKOUT
//                         </CustomButton>
//             </div>
//         </div>
//     );

// };

// export default withRouter(connect(mapStateToProps)(Cart));



import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';

import CartItem from '../CartItem/CartItem';
import { selectCartItems } from '../../redux/cart/cart.selectors';
import { toggleCartHidden } from '../../redux/cart/cart.actions.js';

import {
  CartDropdownContainer,
  CartDropdownButton,
  EmptyMessageContainer,
  CartItemsContainer
} from './CartDropdown.styles';

export const Cart = ({ cartItems, history, dispatch }) => (
  <CartDropdownContainer>
    <CartItemsContainer>
      {cartItems.length ? (
        cartItems.map(cartItem => (
          <CartItem key={cartItem.id} item={cartItem} />
        ))
      ) : (
        <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
      )}
    </CartItemsContainer>
    <CartDropdownButton
      onClick={() => {
        history.push('/checkout');
        dispatch(toggleCartHidden());
      }}
    >
      GO TO CHECKOUT
    </CartDropdownButton>
  </CartDropdownContainer>
);

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
});

export default withRouter(connect(mapStateToProps)(Cart));