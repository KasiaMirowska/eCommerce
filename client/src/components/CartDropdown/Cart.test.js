import React from 'react';
import { shallow } from 'enzyme';
import { Cart } from './Cart';
import CartItem from '../CartItem/CartItem';
import CustomButton from '../CustomButton/CustomButton';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

describe('CartDropdown component', () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn()
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch
    };

    wrapper = shallow(<Cart {...mockProps} />);
  });

  it('should render CartDropdown component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call history.push when button is clicked', () => {
    wrapper.find('CartDropdownButton').simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });

  it('should render an equal number of CartItem components as the cartItems prop', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });

  it('should render EmptyMessageContainer if cartItems is empty', () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch
    };

    const newWrapper = shallow(<Cart {...mockProps} />);
    expect(newWrapper.exists('EmptyMessageContainer')).toBe(true);
  });
});