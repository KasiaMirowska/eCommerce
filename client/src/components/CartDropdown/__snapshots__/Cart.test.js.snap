// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component renders as expected 1`] = `ShallowWrapper {}`;

exports[`CartDropdown component should render CartDropdown component 1`] = `ShallowWrapper {}`;
